### PR TITLE
37737 fixed internal error when replacing with byte-equal parts

### PR DESCRIPTION
**Pull request by priera**
### Changelog category (leave one):
- Bug Fix 

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed error which prevented replacing partitions when the source table had parts equal byte to byte in that partition

Referred in ticket #37737

### CODE REVIEW

File: tests/queries/0_stateless/00626_replace_partition_from_table_zookeeper.sh [hunk 1]
```diff
@@ -57,6 +57,18 @@ $CLICKHOUSE_CLIENT --query="SYSTEM SYNC REPLICA dst_r2;"
 $CLICKHOUSE_CLIENT --query="SELECT count(), sum(d) FROM dst_r1;"
 $CLICKHOUSE_CLIENT --query="SELECT count(), sum(d) FROM dst_r2;"
 
+$CLICKHOUSE_CLIENT --query="SELECT 'REPLACE empty partition with duplicated parts';"
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/00626_replace_partition_from_table_zookeeper.sh:60]: A new test should be added instead of modifying an existing test.

**Comment by alexey-milovidov** [comment]
The changelog entry should explain - what exactly was fixed.

**Comment by priera** [comment]
I'm working in the points listed above. In the meantime, here's a bit of technical description:

When running a `REPLACE PARTITION` from a source table with byte-equal parts the operation fails. The error message is like this: `Coordination::Exception. Coordination::Exception: Transaction failed: Op #3, path: /clickhouse/tables/default/mergetree1_rep/01/blocks/1_replace_from_DDB6B55418392226C72756C87A657CC1. (KEEPER_EXCEPTION)
` 

This happens because several zookeeper commands are being sent with the same `replace_from_...` path. As a solution, just sent one of such commands. The partitions are replaced in the destination table as they are in the original, i.e, no deduplication/optimization happens.

**Comment by priera** [comment]
Hello, I applied both changes: moved the test to a new one and updated the changelog description. Please let me know any other change this PR may require

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit ed41b0d466ccda380a594b7dd5d8e3c8ebf4dea1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53735/ed41b0d466ccda380a594b7dd5d8e3c8ebf4dea1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -7580,13 +7581,15 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
...
+            const bool is_duplicated_part = replaced_parts.contains(hash_hex);
+            replaced_parts.insert(hash_hex);
 
             if (replace)
                 LOG_INFO(log, "Trying to replace {} with hash_hex {}", src_part->name, hash_hex);
             else
                 LOG_INFO(log, "Trying to attach {} with hash_hex {}", src_part->name, hash_hex);
 
-            String block_id_path = replace ? "" : (fs::path(zookeeper_path) / "blocks" / (partition_id + "_replace_from_" + hash_hex));
+            String block_id_path = (replace || is_duplicated_part) ? "" : (fs::path(zookeeper_path) / "blocks" / (partition_id + "_replace_from_" + hash_hex));
```
Comment (CurtizJ) [inline src/Storages/StorageReplicatedMergeTree.cpp:7592]: But why `replace` is `false` even for `REPLACE PARTITION` query? It looks like there is a bug in another place.

**Review by CurtizJ — APPROVED** [review]

**Comment by priera** [comment]
@CurtizJ , thanks for the review and merging. I just noticed that the test I wrote won't fail in case of a regression (originally `ALTER` statement failed, but the retries mechanism used in the test hides that and the test succeeds). I will open an MR in short fixing the test

**Pull request merged**